NullPointerEx in FlorisContentProvider

#### Environment information
- FlorisBoard Version: 0.3.14-beta08 (64)
- Device: LeMobile X52X
- Android: 10 (cn=Q sdk=29)

#### Features enabled
```kt
smartbar = true
suggestions = false
suggestions_clipboard = true
suggestions_next_word = true
glide = true
clipboard_internal = true
clipboard_history = true
```

#### Attached stacktrace files
<details>
<summary>1627586409108.stacktrace</summary>

```
java.lang.NullPointerException
	at dev.patrickgold.florisboard.ime.clip.provider.FlorisContentProvider$Companion.getInstance(FlorisContentProvider.kt:36)
	at dev.patrickgold.florisboard.ime.clip.FlorisClipboardManager$initialize$2.invokeSuspend(FlorisClipboardManager.kt:307)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

```
</details>

<details>
<summary>1627586409159.stacktrace</summary>

```
java.lang.RuntimeException: Exception while trying to handle coroutine exception
	at kotlinx.coroutines.CoroutineExceptionHandlerKt.handlerException(CoroutineExceptionHandler.kt:38)
	at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.handleCoroutineExceptionImpl(CoroutineExceptionHandlerImpl.kt:33)
	at kotlinx.coroutines.CoroutineExceptionHandlerKt.handleCoroutineException(CoroutineExceptionHandler.kt:33)
	at kotlinx.coroutines.StandaloneCoroutine.handleJobException(Builders.common.kt:193)
	at kotlinx.coroutines.JobSupport.finalizeFinishingState(JobSupport.kt:229)
	at kotlinx.coroutines.JobSupport.tryMakeCompletingSlowPath(JobSupport.kt:903)
	at kotlinx.coroutines.JobSupport.tryMakeCompleting(JobSupport.kt:860)
	at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core(JobSupport.kt:825)
	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:111)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
	Suppressed: java.lang.NullPointerException
		at dev.patrickgold.florisboard.ime.clip.provider.FlorisContentProvider$Companion.getInstance(FlorisContentProvider.kt:36)
		at dev.patrickgold.florisboard.ime.clip.FlorisClipboardManager$initialize$2.invokeSuspend(FlorisClipboardManager.kt:307)
		at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
		... 5 more
Caused by: java.lang.IllegalStateException: SharedPreferences in credential encrypted storage are not available until after user is unlocked
	at android.app.ContextImpl.getSharedPreferences(ContextImpl.java:458)
	at android.app.ContextImpl.getSharedPreferences(ContextImpl.java:443)
	at android.content.ContextWrapper.getSharedPreferences(ContextWrapper.java:178)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$Companion.getLastCrashTimestamp(CrashUtility.kt:213)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$Companion.access$getLastCrashTimestamp(CrashUtility.kt:46)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$UncaughtExceptionHandler.uncaughtException(CrashUtility.kt:389)
	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1073)
	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1068)
	at kotlinx.coroutines.android.AndroidExceptionPreHandler.handleException(AndroidExceptionPreHandler.kt:47)
	at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.handleCoroutineExceptionImpl(CoroutineExceptionHandlerImpl.kt:29)
	... 13 more

```
</details>

<details>
<summary>1627586409202.stacktrace</summary>

```
kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for DispatchedContinuation[Dispatchers.IO, Continuation at dev.patrickgold.florisboard.ime.clip.FlorisClipboardManager$initialize$2.invokeSuspend(FlorisClipboardManager.kt)@3af394]. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
	at kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(DispatchedTask.kt:144)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:115)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
Caused by: java.lang.IllegalStateException: SharedPreferences in credential encrypted storage are not available until after user is unlocked
	at android.app.ContextImpl.getSharedPreferences(ContextImpl.java:458)
	at android.app.ContextImpl.getSharedPreferences(ContextImpl.java:443)
	at android.content.ContextWrapper.getSharedPreferences(ContextWrapper.java:178)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$Companion.getLastCrashTimestamp(CrashUtility.kt:213)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$Companion.access$getLastCrashTimestamp(CrashUtility.kt:46)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$UncaughtExceptionHandler.uncaughtException(CrashUtility.kt:389)
	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1073)
	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1068)
	at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.handleCoroutineExceptionImpl(CoroutineExceptionHandlerImpl.kt:33)
	at kotlinx.coroutines.CoroutineExceptionHandlerKt.handleCoroutineException(CoroutineExceptionHandler.kt:33)
	at kotlinx.coroutines.StandaloneCoroutine.handleJobException(Builders.common.kt:193)
	at kotlinx.coroutines.JobSupport.finalizeFinishingState(JobSupport.kt:229)
	at kotlinx.coroutines.JobSupport.tryMakeCompletingSlowPath(JobSupport.kt:903)
	at kotlinx.coroutines.JobSupport.tryMakeCompleting(JobSupport.kt:860)
	at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core(JobSupport.kt:825)
	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:111)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	... 4 more

```
</details>

<details>
<summary>1627586409209.stacktrace</summary>

```
java.lang.RuntimeException: Exception while trying to handle coroutine exception
	at kotlinx.coroutines.CoroutineExceptionHandlerKt.handlerException(CoroutineExceptionHandler.kt:38)
	at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.handleCoroutineExceptionImpl(CoroutineExceptionHandlerImpl.kt:33)
	at kotlinx.coroutines.CoroutineExceptionHandlerKt.handleCoroutineException(CoroutineExceptionHandler.kt:33)
	at kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(DispatchedTask.kt:146)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:115)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
	Suppressed: kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for DispatchedContinuation[Dispatchers.IO, Continuation at dev.patrickgold.florisboard.ime.clip.FlorisClipboardManager$initialize$2.invokeSuspend(FlorisClipboardManager.kt)@3af394]. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
		at kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(DispatchedTask.kt:144)
		... 5 more
	Caused by: java.lang.IllegalStateException: SharedPreferences in credential encrypted storage are not available until after user is unlocked
		at android.app.ContextImpl.getSharedPreferences(ContextImpl.java:458)
		at android.app.ContextImpl.getSharedPreferences(ContextImpl.java:443)
		at android.content.ContextWrapper.getSharedPreferences(ContextWrapper.java:178)
		at dev.patrickgold.florisboard.crashutility.CrashUtility$Companion.getLastCrashTimestamp(CrashUtility.kt:213)
		at dev.patrickgold.florisboard.crashutility.CrashUtility$Companion.access$getLastCrashTimestamp(CrashUtility.kt:46)
		at dev.patrickgold.florisboard.crashutility.CrashUtility$UncaughtExceptionHandler.uncaughtException(CrashUtility.kt:389)
		at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1073)
		at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1068)
		at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.handleCoroutineExceptionImpl(CoroutineExceptionHandlerImpl.kt:33)
		at kotlinx.coroutines.CoroutineExceptionHandlerKt.handleCoroutineException(CoroutineExceptionHandler.kt:33)
		at kotlinx.coroutines.StandaloneCoroutine.handleJobException(Builders.common.kt:193)
		at kotlinx.coroutines.JobSupport.finalizeFinishingState(JobSupport.kt:229)
		at kotlinx.coroutines.JobSupport.tryMakeCompletingSlowPath(JobSupport.kt:903)
		at kotlinx.coroutines.JobSupport.tryMakeCompleting(JobSupport.kt:860)
		at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core(JobSupport.kt:825)
		at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:111)
		at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
		at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
		... 4 more
Caused by: java.lang.IllegalStateException: SharedPreferences in credential encrypted storage are not available until after user is unlocked
	at android.app.ContextImpl.getSharedPreferences(ContextImpl.java:458)
	at android.app.ContextImpl.getSharedPreferences(ContextImpl.java:443)
	at android.content.ContextWrapper.getSharedPreferences(ContextWrapper.java:178)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$Companion.getLastCrashTimestamp(CrashUtility.kt:213)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$Companion.access$getLastCrashTimestamp(CrashUtility.kt:46)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$UncaughtExceptionHandler.uncaughtException(CrashUtility.kt:389)
	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1073)
	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1068)
	at kotlinx.coroutines.android.AndroidExceptionPreHandler.handleException(AndroidExceptionPreHandler.kt:47)
	at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.handleCoroutineExceptionImpl(CoroutineExceptionHandlerImpl.kt:29)
	... 7 more

```
</details>

<details>
<summary>1627586409244.stacktrace</summary>

```
java.lang.IllegalStateException: SharedPreferences in credential encrypted storage are not available until after user is unlocked
	at android.app.ContextImpl.getSharedPreferences(ContextImpl.java:458)
	at android.app.ContextImpl.getSharedPreferences(ContextImpl.java:443)
	at android.content.ContextWrapper.getSharedPreferences(ContextWrapper.java:178)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$Companion.getLastCrashTimestamp(CrashUtility.kt:213)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$Companion.access$getLastCrashTimestamp(CrashUtility.kt:46)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$UncaughtExceptionHandler.uncaughtException(CrashUtility.kt:389)
	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1073)
	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1068)
	at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.handleCoroutineExceptionImpl(CoroutineExceptionHandlerImpl.kt:33)
	at kotlinx.coroutines.CoroutineExceptionHandlerKt.handleCoroutineException(CoroutineExceptionHandler.kt:33)
	at kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(DispatchedTask.kt:146)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:115)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

```
</details>

<details>
<summary>1627586409256.stacktrace</summary>

```
java.lang.IllegalStateException: SharedPreferences in credential encrypted storage are not available until after user is unlocked
	at android.app.ContextImpl.getSharedPreferences(ContextImpl.java:458)
	at android.app.ContextImpl.getSharedPreferences(ContextImpl.java:443)
	at android.content.ContextWrapper.getSharedPreferences(ContextWrapper.java:178)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$Companion.getLastCrashTimestamp(CrashUtility.kt:213)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$Companion.access$getLastCrashTimestamp(CrashUtility.kt:46)
	at dev.patrickgold.florisboard.crashutility.CrashUtility$UncaughtExceptionHandler.uncaughtException(CrashUtility.kt:389)
	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1073)
	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1068)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:574)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

```
</details>

<details>
<summary>1639250186301.stacktrace</summary>

```
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.get(ArrayList.java:437)
	at dev.patrickgold.florisboard.ime.keyboard.LayoutManager.mergeLayouts(LayoutManager.kt:241)
	at dev.patrickgold.florisboard.ime.keyboard.LayoutManager.access$mergeLayouts(LayoutManager.kt:65)
	at dev.patrickgold.florisboard.ime.keyboard.LayoutManager$mergeLayouts$1.invokeSuspend(Unknown Source:19)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

```
</details>

<details>
<summary>1642573860251.stacktrace</summary>

```
java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1441)
	at java.util.HashMap$KeyIterator.next(HashMap.java:1465)
	at dev.patrickgold.florisboard.ime.text.gestures.StatisticalGlideTypingClassifier$Pruner.<init>(StatisticalGlideTypingClassifier.kt:403)
	at dev.patrickgold.florisboard.ime.text.gestures.StatisticalGlideTypingClassifier.initializePruner(StatisticalGlideTypingClassifier.kt:149)
	at dev.patrickgold.florisboard.ime.text.gestures.StatisticalGlideTypingClassifier.setLayout(StatisticalGlideTypingClassifier.kt:117)
	at dev.patrickgold.florisboard.ime.text.gestures.GlideTypingManager.setLayout(GlideTypingManager.kt:67)
	at dev.patrickgold.florisboard.ime.text.keyboard.TextKeyboardLayoutKt.TextKeyboardLayout(TextKeyboardLayout.kt:123)
	at dev.patrickgold.florisboard.ime.text.TextInputLayoutKt.TextInputLayout(TextInputLayout.kt:46)
	at dev.patrickgold.florisboard.ime.text.TextInputLayoutKt$TextInputLayout$2.invoke(Unknown Source:8)
	at dev.patrickgold.florisboard.ime.text.TextInputLayoutKt$TextInputLayout$2.invoke(Unknown Source:8)
	at androidx.compose.runtime.RecomposeScopeImpl.compose(RecomposeScopeImpl.kt:140)
	at androidx.compose.runtime.ComposerImpl.recomposeToGroupEnd(Composer.kt:2158)
	at androidx.compose.runtime.ComposerImpl.skipCurrentGroup(Composer.kt:2404)
	at androidx.compose.runtime.ComposerImpl$doCompose$2$5.invoke(Composer.kt:2585)
	at androidx.compose.runtime.ComposerImpl$doCompose$2$5.invoke(Composer.kt:2571)
	at androidx.compose.runtime.SnapshotStateKt__DerivedStateKt.observeDerivedStateRecalculations(DerivedState.kt:247)
	at androidx.compose.runtime.SnapshotStateKt.observeDerivedStateRecalculations(Unknown Source:1)
	at androidx.compose.runtime.ComposerImpl.doCompose(Composer.kt:2571)
	at androidx.compose.runtime.ComposerImpl.recompose$runtime_release(Composer.kt:2547)
	at androidx.compose.runtime.CompositionImpl.recompose(Composition.kt:620)
	at androidx.compose.runtime.Recomposer.performRecompose(Recomposer.kt:786)
	at androidx.compose.runtime.Recomposer.access$performRecompose(Recomposer.kt:105)
	at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$2.invoke(Recomposer.kt:456)
	at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$2.invoke(Recomposer.kt:425)
	at androidx.compose.ui.platform.AndroidUiFrameClock$withFrameNanos$2$callback$1.doFrame(AndroidUiFrameClock.android.kt:34)
	at androidx.compose.ui.platform.AndroidUiDispatcher.performFrameDispatch(AndroidUiDispatcher.android.kt:109)
	at androidx.compose.ui.platform.AndroidUiDispatcher.access$performFrameDispatch(AndroidUiDispatcher.android.kt:41)
	at androidx.compose.ui.platform.AndroidUiDispatcher$dispatchCallback$1.doFrame(AndroidUiDispatcher.android.kt:69)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:965)
	at android.view.Choreographer.doCallbacks(Choreographer.java:791)
	at android.view.Choreographer.doFrame(Choreographer.java:722)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:952)
	at android.os.Handler.handleCallback(Handler.java:883)
	at android.os.Handler.dispatchMessage(Handler.java:100)
	at android.os.Looper.loop(Looper.java:214)
	at android.app.ActivityThread.main(ActivityThread.java:7356)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:491)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:940)

```
</details>

--- a/app/src/main/java/dev/patrickgold/florisboard/ime/clipboard/ClipboardInputLayout.kt
+++ b/app/src/main/java/dev/patrickgold/florisboard/ime/clipboard/ClipboardInputLayout.kt
@@ -88,7 +88,6 @@ import dev.patrickgold.florisboard.ime.theme.FlorisImeTheme
 import dev.patrickgold.florisboard.ime.theme.FlorisImeUi
 import dev.patrickgold.florisboard.keyboardManager
 import dev.patrickgold.florisboard.snygg.SnyggPropertySet
-import dev.patrickgold.florisboard.snygg.ui.NoContentPadding
 import dev.patrickgold.florisboard.snygg.ui.SnyggSurface
 import dev.patrickgold.florisboard.snygg.ui.snyggBackground
 import dev.patrickgold.florisboard.snygg.ui.snyggBorder
@@ -203,7 +202,6 @@ fun ClipboardInputLayout(
                 .padding(ItemMargin),
             style = style,
             clip = true,
-            contentPadding = if (item.type == ItemType.IMAGE) NoContentPadding else ItemPadding,
             clickAndSemanticsModifier = Modifier.combinedClickable(
                 interactionSource = remember { MutableInteractionSource() },
                 indication = rememberRipple(),
@@ -220,17 +218,36 @@ fun ClipboardInputLayout(
                 val id = ContentUris.parseId(item.uri!!)
                 val file = ClipboardFileStorage.getFileForId(context, id)
                 val bitmap = remember(id) {
-                    BitmapFactory.decodeFile(file.absolutePath).asImageBitmap()
+                    runCatching {
+                        check(file.exists()) { "Unable to resolve image at ${file.absolutePath}" }
+                        val rawBitmap = BitmapFactory.decodeFile(file.absolutePath)
+                        checkNotNull(rawBitmap) { "Unable to decode image at ${file.absolutePath}" }
+                        rawBitmap.asImageBitmap()
+                    }
                 }
-                Image(
-                    modifier = Modifier.fillMaxWidth(),
-                    bitmap = bitmap,
-                    contentDescription = null,
-                    contentScale = ContentScale.FillWidth,
-                )
+                if (bitmap.isSuccess) {
+                    Image(
+                        modifier = Modifier.fillMaxWidth(),
+                        bitmap = bitmap.getOrThrow(),
+                        contentDescription = null,
+                        contentScale = ContentScale.FillWidth,
+                    )
+                } else {
+                    Text(
+                        modifier = Modifier
+                            .fillMaxWidth()
+                            .padding(ItemPadding),
+                        text = bitmap.exceptionOrNull()?.message ?: "Unknown error",
+                        style = TextStyle(textDirection = TextDirection.Ltr),
+                        color = Color.Red,
+                        fontSize = style.fontSize.spSize(),
+                    )
+                }
             } else {
                 Text(
-                    modifier = Modifier.fillMaxWidth(),
+                    modifier = Modifier
+                        .fillMaxWidth()
+                        .padding(ItemPadding),
                     text = item.stringRepresentation(),
                     style = TextStyle(textDirection = TextDirection.ContentOrLtr),
                     color = style.foreground.solidColor(),

--- a/app/src/main/java/dev/patrickgold/florisboard/ime/clipboard/ClipboardManager.kt
+++ b/app/src/main/java/dev/patrickgold/florisboard/ime/clipboard/ClipboardManager.kt
@@ -27,6 +27,7 @@ import dev.patrickgold.florisboard.appContext
 import dev.patrickgold.florisboard.common.android.AndroidClipboardManager
 import dev.patrickgold.florisboard.common.android.AndroidClipboardManager_OnPrimaryClipChangedListener
 import dev.patrickgold.florisboard.common.android.setOrClearPrimaryClip
+import dev.patrickgold.florisboard.common.android.showShortToast
 import dev.patrickgold.florisboard.common.android.systemService
 import dev.patrickgold.florisboard.common.kotlin.tryOrNull
 import dev.patrickgold.florisboard.ime.clipboard.provider.*
@@ -309,7 +310,11 @@ class ClipboardManager(
 
     fun pasteItem(item: ClipboardItem) {
         val activeEditorInstance = FlorisImeService.activeEditorInstance() ?: return
-        activeEditorInstance.commitClipboardItem(item)
+        activeEditorInstance.commitClipboardItem(item).also { result ->
+            if (!result) {
+                appContext.showShortToast("Failed to paste item.")
+            }
+        }
     }
 
     /**

--- a/app/src/main/java/dev/patrickgold/florisboard/ime/core/EditorInstance.kt
+++ b/app/src/main/java/dev/patrickgold/florisboard/ime/core/EditorInstance.kt
@@ -17,6 +17,7 @@
 package dev.patrickgold.florisboard.ime.core
 
 import android.content.ClipDescription
+import android.content.ContentUris
 import android.content.Intent
 import android.inputmethodservice.InputMethodService
 import android.os.SystemClock
@@ -41,6 +42,7 @@ import dev.patrickgold.florisboard.debug.LogTopic
 import dev.patrickgold.florisboard.debug.flogDebug
 import dev.patrickgold.florisboard.debug.flogInfo
 import dev.patrickgold.florisboard.debug.flogWarning
+import dev.patrickgold.florisboard.ime.clipboard.provider.ClipboardFileStorage
 import dev.patrickgold.florisboard.ime.clipboard.provider.ClipboardItem
 import dev.patrickgold.florisboard.ime.clipboard.provider.ItemType
 import dev.patrickgold.florisboard.ime.keyboard.ImeOptions
@@ -398,8 +400,12 @@ class EditorInstance(private val ims: InputMethodService) {
         val mimeTypes = item.mimeTypes
         return when (item.type) {
             ItemType.IMAGE -> {
+                item.uri ?: return false
+                val id = ContentUris.parseId(item.uri)
+                val file = ClipboardFileStorage.getFileForId(ims, id)
+                if (!file.exists()) return false
                 val inputContentInfo = InputContentInfoCompat(
-                    item.uri!!,
+                    item.uri,
                     ClipDescription("clipboard image", mimeTypes),
                     null,
                 )
@@ -672,7 +678,11 @@ class EditorInstance(private val ims: InputMethodService) {
     fun performClipboardPaste(): Boolean {
         isPhantomSpaceActive = false
         wasPhantomSpaceActiveLastUpdate = false
-        return commitClipboardItem(clipboardManager.primaryClip.value)
+        return commitClipboardItem(clipboardManager.primaryClip.value).also { result ->
+            if (!result) {
+                ims.showShortToast("Failed to paste item.")
+            }
+        }
     }
 
     /**

--- a/app/src/main/java/dev/patrickgold/florisboard/ime/core/EditorInstance.kt
+++ b/app/src/main/java/dev/patrickgold/florisboard/ime/core/EditorInstance.kt
@@ -38,6 +38,7 @@ import dev.patrickgold.florisboard.clipboardManager
 import dev.patrickgold.florisboard.common.FlorisLocale
 import dev.patrickgold.florisboard.common.android.AndroidVersion
 import dev.patrickgold.florisboard.common.android.showShortToast
+import dev.patrickgold.florisboard.common.kotlin.tryOrNull
 import dev.patrickgold.florisboard.debug.LogTopic
 import dev.patrickgold.florisboard.debug.flogDebug
 import dev.patrickgold.florisboard.debug.flogInfo
@@ -867,7 +868,7 @@ class EditorInstance(private val ims: InputMethodService) {
         lastReportedComposingBounds = Bounds(-1, -1)
     }
 
-    internal fun normalizeBounds(start: Int, end: Int): Bounds {
+    private fun normalizeBounds(start: Int, end: Int): Bounds {
         return if (start > end) {
             Bounds(end, start)
         } else {
@@ -962,10 +963,12 @@ class EditorInstance(private val ims: InputMethodService) {
          */
         val text: String
             get() {
-                val eiText = cachedInput.rawText
-                val eiStart = (start - cachedInput.offset).coerceIn(0, eiText.length)
-                val eiEnd = (end - cachedInput.offset).coerceIn(0, eiText.length)
-                return if (!isValid || eiEnd - eiStart <= 0) { "" } else { eiText.substring(eiStart, eiEnd) }
+                return tryOrNull {
+                    val eiText = cachedInput.rawText
+                    val eiStart = (start - cachedInput.offset).coerceIn(0, eiText.length)
+                    val eiEnd = (end - cachedInput.offset).coerceIn(0, eiText.length)
+                    if (!isValid || eiEnd - eiStart <= 0) { "" } else { eiText.substring(eiStart, eiEnd) }
+                } ?: ""
             }
 
         val icText: String? get() = when {
@@ -1021,21 +1024,23 @@ class EditorInstance(private val ims: InputMethodService) {
             private set
 
         fun updateText(exText: ExtractedText?) {
-            if (exText == null) {
-                reset()
-                return
-            }
-            val sel = exText.getSelectionBounds()
-            if (selection.bounds != sel) {
-                flogWarning { "Selection from extracted text mismatches from selection state, fixing!" }
-                selection.bounds = sel
-            }
-            if (exText.isPartialChange()) {
-                val (partialStart, partialEnd) = exText.getPartialChangeBounds()
-                rawText.replace(partialStart, partialEnd, exText.getTextStr())
-            } else {
-                rawText.replace(0, rawText.length, exText.getTextStr())
-                offset = exText.startOffset.coerceAtLeast(0)
+            tryOrNull {
+                if (exText == null) {
+                    reset()
+                    return
+                }
+                val sel = exText.getSelectionBounds()
+                if (selection.bounds != sel) {
+                    flogWarning { "Selection from extracted text mismatches from selection state, fixing!" }
+                    selection.bounds = sel
+                }
+                if (exText.isPartialChange()) {
+                    val (partialStart, partialEnd) = exText.getPartialChangeBounds()
+                    rawText.replace(partialStart, partialEnd, exText.getTextStr())
+                } else {
+                    rawText.replace(0, rawText.length, exText.getTextStr())
+                    offset = exText.startOffset.coerceAtLeast(0)
+                }
             }
         }
 

--- a/app/src/main/java/dev/patrickgold/florisboard/ime/text/gestures/StatisticalGlideTypingClassifier.kt
+++ b/app/src/main/java/dev/patrickgold/florisboard/ime/text/gestures/StatisticalGlideTypingClassifier.kt
@@ -587,10 +587,10 @@ class StatisticalGlideTypingClassifier : GlideTypingClassifier {
             return normalizedGesture
         }
 
-        fun getFirstX(): Float = xs[0]
-        fun getFirstY(): Float = ys[0]
-        fun getLastX(): Float = xs[size - 1]
-        fun getLastY(): Float = ys[size - 1]
+        fun getFirstX(): Float = xs.getOrElse(0) { 0f }
+        fun getFirstY(): Float = ys.getOrElse(0) { 0f }
+        fun getLastX(): Float = xs.getOrElse(size - 1) { 0f }
+        fun getLastY(): Float = ys.getOrElse(size - 1) { 0f }
 
         fun getLength(): Float {
             var length = 0f
@@ -609,8 +609,8 @@ class StatisticalGlideTypingClassifier : GlideTypingClassifier {
             this.size = 0
         }
 
-        fun getX(i: Int): Float = xs[i]
-        fun getY(i: Int): Float = ys[i]
+        fun getX(i: Int): Float = xs.getOrElse(i) { 0f }
+        fun getY(i: Int): Float = ys.getOrElse(i) { 0f }
 
         fun clone(): Gesture {
             return Gesture(xs.clone(), ys.clone(), size)

--- a/app/src/main/java/dev/patrickgold/florisboard/ime/text/gestures/StatisticalGlideTypingClassifier.kt
+++ b/app/src/main/java/dev/patrickgold/florisboard/ime/text/gestures/StatisticalGlideTypingClassifier.kt
@@ -276,7 +276,7 @@ class StatisticalGlideTypingClassifier : GlideTypingClassifier {
     ) {
 
         /** A tree that provides fast access to words based on their first and last letter.  */
-        private val wordTree = HashMap<Pair<Int, Int>, ArrayList<String>>()
+        private val wordTree = Collections.synchronizedMap(HashMap<Pair<Int, Int>, ArrayList<String>>())
 
         /**
          * Finds the words whose start and end letter are closest to the start and end points of the
@@ -300,7 +300,7 @@ class StatisticalGlideTypingClassifier : GlideTypingClassifier {
             for (startKey in startKeys) {
                 for (endKey in endKeys) {
                     val keyPair = Pair(startKey, endKey)
-                    val wordsForKeys = wordTree[keyPair]
+                    val wordsForKeys = synchronized(wordTree) { wordTree[keyPair] }
                     if (wordsForKeys != null) {
                         remainingWords.addAll(wordsForKeys)
                     }
@@ -400,10 +400,12 @@ class StatisticalGlideTypingClassifier : GlideTypingClassifier {
         }
 
         init {
-            for (word in words) {
-                val keyPair = getFirstKeyLastKey(word, keysByCharacter)
-                keyPair?.let {
-                    wordTree.getOrPut(keyPair, { arrayListOf() }).add(word)
+            synchronized(wordTree) {
+                for (word in words) {
+                    val keyPair = getFirstKeyLastKey(word, keysByCharacter)
+                    keyPair?.let {
+                        wordTree.getOrPut(keyPair) { arrayListOf() }.add(word)
+                    }
                 }
             }
         }

